### ISSUE TITLE
jetty.server.HttpInput deadlock

### ISSUE THREAD

**Issue by johngmyers** [eclipse/jetty.project#1146]
Jetty 9.3.12 embedded.

The JAX-RS resource hands off the InputStream to a different thread which does reads on it in order to stream the request body to a separate service. Inducing a particular failure in the destination service triggers the deadlock. I haven't tracked down the specific behavior the failure is inducing.

```
Attaching to process ID 8681, please wait...
Debugger attached successfully.
Server compiler detected.
JVM version is 25.72-b15
Deadlock Detection:

Found one Java-level deadlock:
=============================

"http-client-transientstorage-7626":
  waiting to lock Monitor@0x00007f35b86885d8 (Object@0x00000000e7657da8, a org/eclipse/jetty/io/ssl/SslConnection$DecryptedEndPoint),
  which is held by "http-worker-7401 b3381c9a-7b1d-4de4-ac62-40a89aaa992a"
"http-worker-7401 b3381c9a-7b1d-4de4-ac62-40a89aaa992a":
  waiting to lock Monitor@0x00007f35b8e3f128 (Object@0x00000000ec5b6640, a java/util/ArrayDeque),
  which is held by "http-client-transientstorage-7626"

Found a total of 1 deadlock.

....

Thread 28090: (state = BLOCKED)
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.needsFillInterest() @bci=3, line=429 (Compiled frame)
 - org.eclipse.jetty.io.AbstractEndPoint$1.needsFillInterest() @bci=4, line=45 (Compiled frame)
 - org.eclipse.jetty.io.FillInterest.register(org.eclipse.jetty.util.Callback) @bci=209, line=78 (Compiled frame)
 - org.eclipse.jetty.io.AbstractEndPoint.fillInterested(org.eclipse.jetty.util.Callback) @bci=9, line=128 (Compiled frame)
 - org.eclipse.jetty.server.HttpConnection.blockingReadFillInterested() @bci=8, line=554 (Compiled frame)
 - org.eclipse.jetty.server.HttpInputOverHTTP.blockForContent() @bci=18, line=39 (Compiled frame)
 - org.eclipse.jetty.server.HttpInput$1.blockForContent(org.eclipse.jetty.server.HttpInput) @bci=1, line=846 (Compiled frame)
 - org.eclipse.jetty.server.HttpInput.read(byte[], int, int) @bci=241, line=188 (Compiled frame)
 - org.glassfish.jersey.message.internal.EntityInputStream.read(byte[], int, int) @bci=7, line=102 (Compiled frame)
 - java.io.FilterInputStream.read(byte[], int, int) @bci=7, line=133 (Compiled frame)
 - com.proofpoint.anomalytics.attachment.service.StorageService$TimedInputStream.read(byte[], int, int) @bci=28, line=323 (Compiled frame)
 - java.io.FilterInputStream.read(byte[]) @bci=5, line=107 (Compiled frame)
 - com.proofpoint.http.client.InputStreamBodySource$1.read(byte[]) @bci=9, line=56 (Compiled frame)
 - com.proofpoint.http.client.jetty.JettyHttpClient$BodySourceInputStream.read(byte[]) @bci=5, line=975 (Compiled frame)
 - org.eclipse.jetty.client.util.InputStreamContentProvider$InputStreamContentProviderIterator.hasNext() @bci=33, line=176 (Compiled frame)
 - org.eclipse.jetty.client.HttpContent.advance(java.util.Iterator) @bci=1, line=144 (Compiled frame)
 - org.eclipse.jetty.client.HttpContent.advance() @bci=46, line=138 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpSenderOverHTTP$HeadersCallback.<init>(org.eclipse.jetty.client.http.HttpSenderOverHTTP, org.eclipse.jetty.client.HttpExchange, org.eclipse.jetty.client.HttpContent, org.eclipse.jetty.util.Callback) @bci=159, line=189 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(org.eclipse.jetty.client.HttpExchange, org.eclipse.jetty.client.HttpContent, org.eclipse.jetty.util.Callback) @bci=8, line=58 (Compiled frame)
 - org.eclipse.jetty.client.HttpSender.send(org.eclipse.jetty.client.HttpExchange) @bci=174, line=204 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpChannelOverHTTP.send() @bci=14, line=79 (Compiled frame)
 - org.eclipse.jetty.client.HttpConnection.send(org.eclipse.jetty.client.HttpChannel, org.eclipse.jetty.client.HttpExchange) @bci=65, line=205 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(org.eclipse.jetty.client.HttpExchange) @bci=53, line=202 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(org.eclipse.jetty.client.HttpExchange) @bci=5, line=82 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(org.eclipse.jetty.client.http.HttpConnectionOverHTTP, org.eclipse.jetty.client.HttpExchange) @bci=2, line=37 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(org.eclipse.jetty.client.api.Connection, org.eclipse.jetty.client.HttpExchange) @bci=6, line=27 (Compiled frame)
 - org.eclipse.jetty.client.PoolingHttpDestination.process(org.eclipse.jetty.client.api.Connection) @bci=204, line=165 (Compiled frame)
 - org.eclipse.jetty.client.PoolingHttpDestination.process() @bci=14, line=113 (Compiled frame)
 - org.eclipse.jetty.client.PoolingHttpDestination.send() @bci=14, line=97 (Compiled frame)
 - org.eclipse.jetty.client.PoolingHttpDestination.succeeded() @bci=1, line=84 (Compiled frame)
 - org.eclipse.jetty.client.DuplexConnectionPool.proceed() @bci=4, line=171 (Compiled frame)
 - org.eclipse.jetty.client.DuplexConnectionPool$1.succeeded(org.eclipse.jetty.client.api.Connection) @bci=64, line=148 (Compiled frame)
 - org.eclipse.jetty.client.DuplexConnectionPool$1.succeeded(java.lang.Object) @bci=5, line=139 (Compiled frame)
 - org.eclipse.jetty.util.Promise$Wrapper.succeeded(java.lang.Object) @bci=5, line=78 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onOpen() @bci=13, line=90 (Compiled frame)
 - org.eclipse.jetty.io.ssl.SslConnection.onOpen() @bci=11, line=162 (Compiled frame)
 - org.eclipse.jetty.io.SelectorManager.connectionOpened(org.eclipse.jetty.io.Connection) @bci=1, line=324 (Compiled frame)
 - org.eclipse.jetty.io.ManagedSelector.createEndPoint(java.nio.channels.SocketChannel, java.nio.channels.SelectionKey) @bci=54, line=425 (Compiled frame)
 - org.eclipse.jetty.io.ManagedSelector.access$1600(org.eclipse.jetty.io.ManagedSelector, java.nio.channels.SocketChannel, java.nio.channels.SelectionKey) @bci=3, line=56 (Compiled frame)
 - org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run() @bci=12, line=605 (Compiled frame)
 - org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume() @bci=401, line=303 (Compiled frame)
 - org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume() @bci=28, line=148 (Compiled frame)
 - org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run() @bci=148, line=136 (Compiled frame)
 - org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(java.lang.Runnable) @bci=1, line=671 (Compiled frame)
 - org.eclipse.jetty.util.thread.QueuedThreadPool$2.run() @bci=104, line=589 (Compiled frame)
 - java.lang.Thread.run() @bci=11, line=745 (Compiled frame)

...

Thread 26927: (state = BLOCKED)
 - org.eclipse.jetty.server.HttpInput.failed(java.lang.Throwable) @bci=9, line=628 (Interpreted frame)
 - org.eclipse.jetty.server.HttpConnection$BlockingReadCallback.failed(java.lang.Throwable) @bci=8, line=605 (Interpreted frame)
 - org.eclipse.jetty.io.FillInterest.onFail(java.lang.Throwable) @bci=29, line=125 (Interpreted frame)
 - org.eclipse.jetty.io.AbstractEndPoint.close(java.lang.Throwable) @bci=18, line=121 (Compiled frame)
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.close(java.lang.Throwable) @bci=18, line=974 (Compiled frame)
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.flush(java.nio.ByteBuffer[]) @bci=1041, line=894 (Interpreted frame)
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.shutdownOutput() @bci=104, line=932 (Compiled frame)
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.close() @bci=1, line=964 (Compiled frame)
 - org.eclipse.jetty.io.AbstractConnection.close() @bci=4, line=220 (Compiled frame)
 - org.eclipse.jetty.server.HttpConnection.onFillable() @bci=113, line=244 (Compiled frame)
 - org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded() @bci=4, line=273 (Compiled frame)
 - org.eclipse.jetty.io.FillInterest.fillable() @bci=61, line=95 (Compiled frame)
 - org.eclipse.jetty.io.ssl.SslConnection.onFillable() @bci=56, line=202 (Compiled frame)
 - org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded() @bci=4, line=273 (Compiled frame)
 - org.eclipse.jetty.io.FillInterest.fillable() @bci=61, line=95 (Compiled frame)
 - org.eclipse.jetty.io.SelectChannelEndPoint$2.run() @bci=7, line=93 (Compiled frame)
 - org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(java.lang.Runnable) @bci=1, line=671 (Compiled frame)
 - com.proofpoint.http.server.HttpServer$1.runJob(java.lang.Runnable) @bci=13, line=141 (Compiled frame)
 - org.eclipse.jetty.util.thread.QueuedThreadPool$2.run() @bci=104, line=589 (Compiled frame)
 - java.lang.Thread.run() @bci=11, line=745 (Compiled frame)

```

**Comment by sbordet** [comment]
@johngmyers the deadlock report seems to refer to client threads, but the one of the stack traces refer to the server, so it's not clear who deadlocks who.

Can you please take a full thread dump with locking information when this happens ?

I don't know what tool you used to take the thread dumps, but if you use `ctrl+\` or `jstack`, they report locking information, for example:

```
   java.lang.Thread.State: RUNNABLE
	at java.io.FileInputStream.readBytes(Native Method)
	at java.io.FileInputStream.read(FileInputStream.java:255)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:284)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
	- locked <0x000000064079a558> (a java.lang.UNIXProcess$ProcessPipeInputStream)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	- locked <0x000000064079a5e0> (a com.intellij.util.io.BaseInputStreamReader)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.Reader.read(Reader.java:140)
```

I am interested in the `- locked ...` lines.

**Comment by johngmyers** [comment]
My analysis:

org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.needsFillInterest() attempts to lock the DecryptedEndPoint while org.eclipse.jetty.server.HttpInput.read(byte[], int, int) is holding a lock on the HttpInput._inputQ

org.eclipse.jetty.server.HttpInput.failed(java.lang.Throwable) attempts to lock the HttpInput._inputQ while org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.shutdownOutput() is holding a lock on the DecryptedEndPoint.

The deadlock is between the consumer trying to register interest and the producer trying to produce an error state transition.

As I mentioned in the description, the JAX-RS resource hands off the InputStream to a different thread which does reads on it in order to stream the request body to a separate service. The client's InputStreamContentProvider is what is reading the server's request body InputStream. The deadlock is entirely in the server code.

The thread dump was taken with jstack -F. JMX is disabled because it is so ridiculously difficult to secure. CTRL-\ would log the stacks to /dev/null, so the only way to get those would be to intercept the write() calls with strace.

**Comment by gregw** [comment]
I agree there is definitely a deadlock there.

Typically the approach we have taken is that the SslConnection is passive from below and active from above.  Ie that means it does not act on events coming up from TCP, but rather wakes up a thread that was blocked on the IO to run the SSL conversation.    In this case an error has occurred and we are obviously trying to be active from below and thus deadlocking.

So two issues need to be investigated:

1. Can we handle the errors from below by a wakeup rather than explicit close?
2. Why are we noticing the errors in the first place?  If we were not read interested, then we should not be seeing these events?  If we were read interested, then why are we trying to become read interested?

**Comment by johngmyers** [comment]
The consumer is repeatedly calling org.eclipse.jetty.server.HttpInput.read(byte[], int, int), possibly with significant time between calls.

**Comment by gregw** [comment]
So expanding on what the read thread should be doing:

- read is called, _inputQ is locked, _state.blockForContent is called
  - HttpInputOverHTTP.blockForContent is the impl used, which calls HttpConnection.blockingReadFillInterested
    - this calls `DecryptedEndpoint.fillInterested(_blockingReadCallback)`, which if called back is simply a notify on the inputQ
      - the registration of fill interest eventually calls `DecryptedEndPoint.needsFillInterest`, which is blocked trying to lock the DecryptedEndPoint - if it had got that lock it would have eventually called ensureFillIntereted, but it never go there. 

The DEP is locked because a ReadCallback is already in progress?!?! Why?? somebody must have already called `fillInterested` on the SSLConnection?  The only thing that should do that is ensureFillInterested?

Ah, which can be called from onIncompleteFlush!  So it may be that another thread is writing and has gotten into that Ssl state where it needs to read more before it can flush?!??! Does that sound like your app?

It is also called from shutdownOutput, but nothing should be calling that directly (not without first being onFillable first).

So currently I'm thinking it is a race between an outgoing write and the read?   

Does your app write while reading?

**Comment by gregw** [comment]
Manual annotation of stack trace for reference:

```
Thread 28090: (state = BLOCKED)
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.needsFillInterest() @bci=3, line=429 (Compiled frame)
    - locking DecryptedEndPoint
 - org.eclipse.jetty.io.AbstractEndPoint$1.needsFillInterest() @bci=4, line=45 (Compiled frame)
 - org.eclipse.jetty.io.FillInterest.register(org.eclipse.jetty.util.Callback) @bci=209, line=78 (Compiled frame)
 - org.eclipse.jetty.io.AbstractEndPoint.fillInterested(org.eclipse.jetty.util.Callback) @bci=9, line=128 (Compiled frame)
 - org.eclipse.jetty.server.HttpConnection.blockingReadFillInterested() @bci=8, line=554 (Compiled frame)
 - org.eclipse.jetty.server.HttpInputOverHTTP.blockForContent() @bci=18, line=39 (Compiled frame)
 - org.eclipse.jetty.server.HttpInput$1.blockForContent(org.eclipse.jetty.server.HttpInput) @bci=1, line=846 (Compiled frame)
 - org.eclipse.jetty.server.HttpInput.read(byte[], int, int) @bci=241, line=188 (Compiled frame)
 - org.glassfish.jersey.message.internal.EntityInputStream.read(byte[], int, int) @bci=7, line=102 (Compiled frame)
 - java.io.FilterInputStream.read(byte[], int, int) @bci=7, line=133 (Compiled frame)
 - com.proofpoint.anomalytics.attachment.service.StorageService$TimedInputStream.read(byte[], int, int) @bci=28, line=323 (Compiled frame)
 - java.io.FilterInputStream.read(byte[]) @bci=5, line=107 (Compiled frame)
 - com.proofpoint.http.client.InputStreamBodySource$1.read(byte[]) @bci=9, line=56 (Compiled frame)
   - locked _inputQ
 - com.proofpoint.http.client.jetty.JettyHttpClient$BodySourceInputStream.read(byte[]) @bci=5, line=975 (Compiled frame)
 - org.eclipse.jetty.client.util.InputStreamContentProvider$InputStreamContentProviderIterator.hasNext() @bci=33, line=176 (Compiled frame)
 - org.eclipse.jetty.client.HttpContent.advance(java.util.Iterator) @bci=1, line=144 (Compiled frame)
 - org.eclipse.jetty.client.HttpContent.advance() @bci=46, line=138 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpSenderOverHTTP$HeadersCallback.<init>(org.eclipse.jetty.client.http.HttpSenderOverHTTP, org.eclipse.jetty.client.HttpExchange, org.eclipse.jetty.client.HttpContent, org.eclipse.jetty.util.Callback) @bci=159, line=189 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(org.eclipse.jetty.client.HttpExchange, org.eclipse.jetty.client.HttpContent, org.eclipse.jetty.util.Callback) @bci=8, line=58 (Compiled frame)
 - org.eclipse.jetty.client.HttpSender.send(org.eclipse.jetty.client.HttpExchange) @bci=174, line=204 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpChannelOverHTTP.send() @bci=14, line=79 (Compiled frame)
 - org.eclipse.jetty.client.HttpConnection.send(org.eclipse.jetty.client.HttpChannel, org.eclipse.jetty.client.HttpExchange) @bci=65, line=205 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(org.eclipse.jetty.client.HttpExchange) @bci=53, line=202 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(org.eclipse.jetty.client.HttpExchange) @bci=5, line=82 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(org.eclipse.jetty.client.http.HttpConnectionOverHTTP, org.eclipse.jetty.client.HttpExchange) @bci=2, line=37 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(org.eclipse.jetty.client.api.Connection, org.eclipse.jetty.client.HttpExchange) @bci=6, line=27 (Compiled frame)
 - org.eclipse.jetty.client.PoolingHttpDestination.process(org.eclipse.jetty.client.api.Connection) @bci=204, line=165 (Compiled frame)
 - org.eclipse.jetty.client.PoolingHttpDestination.process() @bci=14, line=113 (Compiled frame)
 - org.eclipse.jetty.client.PoolingHttpDestination.send() @bci=14, line=97 (Compiled frame)
 - org.eclipse.jetty.client.PoolingHttpDestination.succeeded() @bci=1, line=84 (Compiled frame)
 - org.eclipse.jetty.client.DuplexConnectionPool.proceed() @bci=4, line=171 (Compiled frame)
 - org.eclipse.jetty.client.DuplexConnectionPool$1.succeeded(org.eclipse.jetty.client.api.Connection) @bci=64, line=148 (Compiled frame)
 - org.eclipse.jetty.client.DuplexConnectionPool$1.succeeded(java.lang.Object) @bci=5, line=139 (Compiled frame)
 - org.eclipse.jetty.util.Promise$Wrapper.succeeded(java.lang.Object) @bci=5, line=78 (Compiled frame)
 - org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onOpen() @bci=13, line=90 (Compiled frame)
 - org.eclipse.jetty.io.ssl.SslConnection.onOpen() @bci=11, line=162 (Compiled frame)
 - org.eclipse.jetty.io.SelectorManager.connectionOpened(org.eclipse.jetty.io.Connection) @bci=1, line=324 (Compiled frame)
 - org.eclipse.jetty.io.ManagedSelector.createEndPoint(java.nio.channels.SocketChannel, java.nio.channels.SelectionKey) @bci=54, line=425 (Compiled frame)
 - org.eclipse.jetty.io.ManagedSelector.access$1600(org.eclipse.jetty.io.ManagedSelector, java.nio.channels.SocketChannel, java.nio.channels.SelectionKey) @bci=3, line=56 (Compiled frame)
 - org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run() @bci=12, line=605 (Compiled frame)
 - org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume() @bci=401, line=303 (Compiled frame)
 - org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume() @bci=28, line=148 (Compiled frame)
 - org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run() @bci=148, line=136 (Compiled frame)
 - org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(java.lang.Runnable) @bci=1, line=671 (Compiled frame)
 - org.eclipse.jetty.util.thread.QueuedThreadPool$2.run() @bci=104, line=589 (Compiled frame)
 - java.lang.Thread.run() @bci=11, line=745 (Compiled frame)

...

Thread 26927: (state = BLOCKED)
 - org.eclipse.jetty.server.HttpInput.failed(java.lang.Throwable) @bci=9, line=628 (Interpreted frame)
   - locking _inputQ
 - org.eclipse.jetty.server.HttpConnection$BlockingReadCallback.failed(java.lang.Throwable) @bci=8, line=605 (Interpreted frame)
 - org.eclipse.jetty.io.FillInterest.onFail(java.lang.Throwable) @bci=29, line=125 (Interpreted frame)
 - org.eclipse.jetty.io.AbstractEndPoint.close(java.lang.Throwable) @bci=18, line=121 (Compiled frame)
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.close(java.lang.Throwable) @bci=18, line=974 (Compiled frame)
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.flush(java.nio.ByteBuffer[]) @bci=1041, line=894 (Interpreted frame)
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.shutdownOutput() @bci=104, line=932 (Compiled frame)
   - locked DecryptedEndPoint
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.close() @bci=1, line=964 (Compiled frame)
 - org.eclipse.jetty.io.AbstractConnection.close() @bci=4, line=220 (Compiled frame)
 - org.eclipse.jetty.server.HttpConnection.onFillable() @bci=113, line=244 (Compiled frame)
 - org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded() @bci=4, line=273 (Compiled frame)
 - org.eclipse.jetty.io.FillInterest.fillable() @bci=61, line=95 (Compiled frame)
 - org.eclipse.jetty.io.ssl.SslConnection.onFillable() @bci=56, line=202 (Compiled frame)
 - org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded() @bci=4, line=273 (Compiled frame)
 - org.eclipse.jetty.io.FillInterest.fillable() @bci=61, line=95 (Compiled frame)
 - org.eclipse.jetty.io.SelectChannelEndPoint$2.run() @bci=7, line=93 (Compiled frame)
 - org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(java.lang.Runnable) @bci=1, line=671 (Compiled frame)
 - com.proofpoint.http.server.HttpServer$1.runJob(java.lang.Runnable) @bci=13, line=141 (Compiled frame)
 - org.eclipse.jetty.util.thread.QueuedThreadPool$2.run() @bci=104, line=589 (Compiled frame)
 - java.lang.Thread.run() @bci=11, line=745 (Compiled frame)
```

**Comment by gregw** [comment]
I've not yet reproduced, as it will take a moderately complex unit test to capture this.  However I think I can see a specific fix - update shutdownOutput() to not hold the lock on DecryptedEndPoint when flushing or closing.

However, I think a wider fix is needed, as the lock on DecryptedEndPoint looks to be held too widely and really needs to be reviewed and limited to just the state changes on the endpoint and not the actual operations, which may result in reentrant calls.

**Comment by johngmyers** [comment]
I'm still getting deadlocks using a 9.3.0-SNAPSHOT build of Jetty.

My analysis shows it to likely be due to org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.flush(java.nio.ByteBuffer[] being a synchronized method.

```
Found one Java-level deadlock:
=============================

"http-client-transientstorage-974":
  waiting to lock Monitor@0x00007fae449c3538 (Object@0x00000000e57768c0, a org/eclipse/jetty/io/ssl/SslConnection$DecryptedEndPoint),
  which is held by "http-worker-1047 9c435623-5908-4a29-9041-e8ceef86b171"
"http-worker-1047 9c435623-5908-4a29-9041-e8ceef86b171":
  waiting to lock Monitor@0x00007fae3060c078 (Object@0x00000000e5acb2e0, a java/util/ArrayDeque),
  which is held by "http-client-transientstorage-974"


 - org.eclipse.jetty.server.HttpInput.failed(java.lang.Throwable) @bci=9, line=628 
 - org.eclipse.jetty.server.HttpConnection$BlockingReadCallback.failed(java.lang.Throwable) @bci=8, line=605 
 - org.eclipse.jetty.io.FillInterest.onFail(java.lang.Throwable) @bci=29, line=140 
 - org.eclipse.jetty.io.AbstractEndPoint.close(java.lang.Throwable) @bci=18, line=121 
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.close(java.lang.Throwable) @bci=18, line=1021 
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.flush(java.nio.ByteBuffer[]) @bci=1049, line=925 
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.shutdownOutput() @bci=147, line=976 
 - org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.close() @bci=1, line=1011 
 - org.eclipse.jetty.io.AbstractConnection.close() @bci=4, line=230 
 - org.eclipse.jetty.server.HttpConnection.onFillable() @bci=113, line=244 
 - org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded() @bci=4, line=283 
 - org.eclipse.jetty.io.FillInterest.fillable() @bci=61, line=110 
 - org.eclipse.jetty.io.ssl.SslConnection.onFillable() @bci=56, line=223 
 - org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded() @bci=4, line=283 
 - org.eclipse.jetty.io.FillInterest.fillable() @bci=61, line=110 
 - org.eclipse.jetty.io.SelectChannelEndPoint$2.run() @bci=7, line=93 
 - org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume() @bci=401, line=303 
 - org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume() @bci=28, line=148 
 - org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run() @bci=148, line=136 
 - org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(java.lang.Runnable) @bci=1, line=671 
 - com.proofpoint.http.server.HttpServer$1.runJob(java.lang.Runnable) @bci=13, line=141 
 - org.eclipse.jetty.util.thread.QueuedThreadPool$2.run() @bci=104, line=589 
 - java.lang.Thread.run() @bci=11, line=745 

```

**Comment by sbordet** [comment]
Can you specify what commit ID are you using ?

I believe we have fixed those deadlocks, but perhaps not all ?

Do you have locking information on the stack trace ?

**Comment by johngmyers** [comment]
I pulled from your snapshot repository, but Maven doesn't make tracking those details easy. I know I have af5d27c.

The head of jetty-9.3.x shows that org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.flush(java.nio.ByteBuffer[]) is a synchronized method, causing it to hold the DecryptedEndPoint.

I do not have locking information on the stack trace.

**Comment by sbordet** [comment]
@johngmyers I think I fixed this one as well. Are you able to try soon ? We're about to release 9.3.15, so would be great if you can confirm this fix solves your issue. Thanks !

**Comment by johngmyers** [comment]
The good news is that our latest test run did not deadlock.

The bad news is that a large number of threads are blocked on different ArrayDeque objects from org.eclipse.jetty.server.HttpInput.blockForContent(HttpInput.java:398), but all the corresponding server connections are long gone.

But that's a different bug. I will have to do another run with more instrumentation in order to get a Jetty dump to report that one properly.

**Comment by sbordet** [comment]
@johngmyers dang! Please open another issue for the blocked threads.

**Comment by joakime** [comment]
This issue seems satisfied, closing ...
